lib/miq_cockpit.rb

```diff
@@ -0,0 +1,133 @@
+require 'uri'
+require 'awesome_spawn'
+module MiqCockpit
+  class WS
+    DEFAULT_PORT = 9000
+
+    attr_accessor(:config_dir)
+
+    def self.url(opts, default_host)
+      opts ||= {}
+
+      # running somewhere else
+      if opts[:external_url]
+        url = URI.parse(opts[:external_url])
+      else
+        url = URI::HTTP.build(:host   => default_host,
+                              :scheme => opts[:ssl] ? "https" : "http",
+                              :port   => opts[:port] || DEFAULT_PORT)
+      end
+
+      # Force all paths to be the same
+      url.path = MiqCockpit::ApacheConfig.url_root
+      url
+    end
+
+    def initialize(opts = {})
+      @opts = opts || {}
+      @config_dir = File.join(File.dirname(__FILE__), '../config')
+    end
+
+    def command(address)
+      args = { :port => @opts[:port] || DEFAULT_PORT.to_s }
+      if address
+        args[:address] = address
+      end
+      args[:'no-tls'] = nil unless @opts[:ssl]
+
+      AwesomeSpawn.build_command_line('/usr/libexec/cockpit-ws', args)
+    end
+
+    def save_config
+      Dir.mkdir(@config_dir) unless File.exist?(@config_dir)
+      Dir.mkdir(File.join(@config_dir, "cockpit")) unless File.exist?(File.join(@config_dir, "cockpit"))
+      fname = File.join(@config_dir, "cockpit", "cockpit.conf")
+      update_config
+      File.open(fname, "w") { |f| f.write @config }
+      if @opts[:ssl]
+        setup_ssl
+      end
+    end
+
+    def update_config
+      title = "ManageIQ Cockpit"
+
+      login_command = File.expand_path(File.join(File.dirname(__FILE__), '..', 'tools', 'cockpit', 'cockpit-auth-miq'))
+      miq_url = @opts[:miq_url]
+      url_root = MiqCockpit::ApacheConfig::URL_ROOT
+      @config = <<-END_OF_CONFIG
+[Webservice]
+LoginTitle = #{title}
+UrlRoot = #{url_root}
+
+[Negotiate]
+Action = none
+
+[Basic]
+Action = none
+
+[Bearer]
+Action = remote-login-ssh
+
+[SSH-Login]
+command = #{login_command}
+authFD=10
+
+[OAuth]
+Url = #{miq_url}/dashboard/cockpit_redirect
+
+END_OF_CONFIG
+      @config
+    end
+
+    def setup_ssl
+      dir = File.join(@config_dir, "cockpit", "ws-certs.d")
```

Comment:
I'm seeing a pattern here.  You might want to have a cockpit_dir method that returns `File.join(@config_dir, "cockpit")`  (or even better, you might want to use Pathname objects here, which are much more friendly for building File path strings.